File: linked_list/141_ll_cycle.cpp
// 141. Linked List Cycle
#include <iostream>

#include "LinkedList.h"

bool hasCycle(ListNode* head) {
    // Floyd's cycle detection algorithm
    // also called Tortoise and Hare algorithm
    ListNode* slow = head;
    ListNode* fast = head;

    while (fast != nullptr && fast->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;

        // IMPORTANT: Check for equality only after advancing the pointers.
        // slow and fast start at the same node (head), so checking before
        // movement would always return true even when no cycle exists.

        if (slow == fast) {
            return true;
        }
    }

    return false;
}

int main() {
    ListNode* head = nullptr;
    ListNode* tail = nullptr;

    append_node(10, &head, &tail);
    ListNode* n2 = append_node(20, &head, &tail);
    append_node(30, &head, &tail);
    print_list(head);

    std::cout << hasCycle(head) << std::endl;

    // creating a cycle
    ListNode* n4 = append_node(40, &head, &tail);
    n4->next = n2;
    std::cout << hasCycle(head) << std::endl;

    return 0;
}

File: linked_list/141_ll_cycle.py
# 141. Linked List Cycle
from include.linked_list import ListNode
from typing import Optional


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        slow = head
        fast = head

        while fast is not None and fast.next is not None:
            slow = slow.next
            fast = fast.next.next

            if slow is fast:
                return True

        return False

    def hasCycle2(self, head: Optional[ListNode]) -> bool:
        seen: set[ListNode] = set()
        curr = head

        while curr is not None:
            if curr in seen:
                return True
            seen.add(curr)
            curr = curr.next

        return False
